src/input/json_to_storage.cc

```diff
@@ -121,78 +181,381 @@ JSONPath JSONPath::find_ref_node(const string& ref_address)
+}
+
+
+bool PathYAML::get_bool_value() const {
+	if (head()->IsScalar()) {
+		try {
+			return head()->as<bool>();
+		} catch (YAML::Exception) {
+	        THROW( JSONToStorage::ExcInputError() << JSONToStorage::EI_Specification("The value should be 'YAML bool', but we found: ")
+	                << JSONToStorage::EI_ErrorAddress(this->as_string()) << JSONToStorage::EI_JSON_Type( get_node_type() )
+	             );
+		}
+	} else {
+        THROW( JSONToStorage::ExcInputError() << JSONToStorage::EI_Specification("The value should be 'YAML bool', but we found: ")
+                << JSONToStorage::EI_ErrorAddress(this->as_string()) << JSONToStorage::EI_JSON_Type( get_node_type() )
+             );
+	}
+	return false;
 }
 
 
+std::int64_t PathYAML::get_int_value() const {
+	if (head()->IsScalar()) {
+		try {
+			return head()->as<std::int64_t>();
+		} catch (YAML::Exception) {
+	        THROW( JSONToStorage::ExcInputError() << JSONToStorage::EI_Specification("The value should be 'YAML int', but we found: ")
+	                << JSONToStorage::EI_ErrorAddress(this->as_string()) << JSONToStorage::EI_JSON_Type( get_node_type() )
+	             );
+		}
+	} else {
+        THROW( JSONToStorage::ExcInputError() << JSONToStorage::EI_Specification("The value should be 'YAML int', but we found: ")
+                << JSONToStorage::EI_ErrorAddress(this->as_string()) << JSONToStorage::EI_JSON_Type( get_node_type() )
+             );
+	}
+	return 0;
+}
+
+
+double PathYAML::get_double_value() const {
+	if (head()->IsScalar()) {
+		try {
+			return head()->as<double>();
+		} catch (YAML::Exception) {
+	        THROW( JSONToStorage::ExcInputError() << JSONToStorage::EI_Specification("The value should be 'YAML double', but we found: ")
+	                << JSONToStorage::EI_ErrorAddress(this->as_string()) << JSONToStorage::EI_JSON_Type( get_node_type() )
+	             );
+		}
+	} else {
+        THROW( JSONToStorage::ExcInputError() << JSONToStorage::EI_Specification("The value should be 'YAML double', but we found: ")
+                << JSONToStorage::EI_ErrorAddress(this->as_string()) << JSONToStorage::EI_JSON_Type( get_node_type() )
+             );
+	}
+	return 0.0;
+}
+
+
+std::string PathYAML::get_string_value() const {
+	if (head()->IsScalar()) {
+		try {
+			return head()->as<std::string>();
+		} catch (YAML::Exception) {
+	        THROW( JSONToStorage::ExcInputError() << JSONToStorage::EI_Specification("The value should be 'YAML string', but we found: ")
+	                << JSONToStorage::EI_ErrorAddress(this->as_string()) << JSONToStorage::EI_JSON_Type( get_node_type() )
+	             );
+		}
+	} else {
+        THROW( JSONToStorage::ExcInputError() << JSONToStorage::EI_Specification("The value should be 'YAML string', but we found: ")
+                << JSONToStorage::EI_ErrorAddress(this->as_string()) << JSONToStorage::EI_JSON_Type( get_node_type() )
+             );
+	}
+	return "";
+}
+
+
+std::string PathYAML::get_node_type() const {
```

Comment:
1) vector of type names declared in PathBase 
2) fill type names vector in PathJSON and PathYAML, use common position for null, map, sequence;
leave Scalar possition oempty for JSON and vice versa for individaul scalar types in YAML.
3) move get_map_name, get_scalar_name implementation to PathBase, just returning the item of type names vector
4) rename those methods to get_record_reader_type, get_array_reader_type
5) in PathYAML::get_node_type return appropriate items of type names vector.